lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs

```diff
@@ -72,30 +72,92 @@ pub struct TeleportRdpdrBackend {
         Self {
             active_device_ids: vec![smartcard_device_id],
             client_handle,
+            contexts: Contexts::new(),
         }
     }
 
     fn get_scard_device_id(&self) -> PduResult<u32> {
-        if self.active_device_ids.len() == 0 {
-            return Err(other_err!(
+        if self.active_device_ids.is_empty() {
+            return Err(custom_err!(
                 "TeleportRdpdrBackend::get_scard_device_id",
-                "no active devices",
+                TeleportRdpdrBackendError("no active devices".to_string())
             ));
         }
         Ok(self.active_device_ids[0])
     }
+
+    fn handle_access_started(
+        &mut self,
+        req: DeviceControlRequest<ScardIoCtlCode>,
+        _call: ScardAccessStartedEventCall,
+    ) -> PduResult<()> {
+        let scard_device_id = self.get_scard_device_id()?;
+        if req.header.device_id != scard_device_id {
+            return Err(custom_err!(
+                "TeleportRdpdrBackend::handle_scard_access_started_event_call",
+                TeleportRdpdrBackendError(
+                    format!(
+                        "got ScardAccessStartedEventCall for unknown device_id [{}], expected [{}]",
+                        req.header.device_id, scard_device_id
+                    )
+                    .to_string()
+                ),
+            ));
+        }
+
+        self.write_rdpdr_dev_ctl_resp(req, Box::new(LongReturn::new(ReturnCode::Success)))
+    }
+
+    fn handle_establish_context(
+        &mut self,
+        req: DeviceControlRequest<ScardIoCtlCode>,
+        _call: EstablishContextCall,
+    ) -> PduResult<()> {
+        let ctx = self.contexts.establish();
+
+        self.write_rdpdr_dev_ctl_resp(
+            req,
+            Box::new(EstablishContextReturn::new(ReturnCode::Success, ctx)),
+        )
+    }
+
+    fn write_rdpdr_dev_ctl_resp(
+        &mut self,
+        req: DeviceControlRequest<ScardIoCtlCode>,
+        resp: Box<dyn rpce::Encode>,
+    ) -> PduResult<()> {
+        let resp = DeviceControlResponse::new(req, NtStatus::Success, resp);
+        self.client_handle
+            .blocking_send(ClientFunction::WriteRdpdr(RdpdrPdu::DeviceControlResponse(
+                resp,
+            )))
+            .map_err(|e| {
+                custom_err!(
+                    "write_rdpdr_dev_ctl_resp",
+                    // Due to a long chain of trait dependencies in IronRDP that are impractical to unwind at this point,
+                    // we can't put _e in the source field of the error because it isn't Sync (because ClientFunction itself
+                    // isn't sync). We compromise here by just wrapping its Debug output in a TeleportRdpdrBackendError.
+                    TeleportRdpdrBackendError(format!("{:?}", e))
+                )
+            })
+    }
 }
 
 impl RdpdrBackend for TeleportRdpdrBackend {
     fn handle_server_device_announce_response(
-        &self,
+        &mut self,
```

Comment:
Why do we need a mutable ref here?

Reply:
The narrow reason is that that's what the [`RdpdrBackend` over in IronRDP](https://github.com/Devolutions/IronRDP/blob/master/crates/ironrdp-rdpdr/src/backend/mod.rs#L12) calls for. 

The broader reason I designed the trait that way is that it gives the `RdpdrBackend` implementer maximal flexibility, even though we don't need to mutate anything on `ServerDeviceAnnounceResponse`, maybe we will in the future, or IronRDP's native Windows/Linux/MacOS/Browser implementation of `RdpdrBackend` needs to.

(The "maximal flexibility" has a caveat wrt to the potential for multi-threaded implementations, we're currently leaning against encouraging that for some reasons Benoit lists in https://github.com/Devolutions/IronRDP/pull/208#discussion_r1349387928).